service/subresource_destroyer.go

```diff
@@ -0,0 +1,158 @@
+// resource in an inconsistent state.
+//
+// Plus, it's not the client's fault that we weren't able to delete the authentications, so by taking this approach
+// the clients don't keep an "unremovable" resource because of some failure in deleting the authentications.
+//
+// Finally, those authentications are safely encrypted so they can stay in their datastores until we manually remove
+// them.
+func DeleteCascade(tenantId *int64, resourceType string, resourceId int64, headers []kafka.Header) error {
+	authenticationsDao := dao.GetAuthenticationDao(tenantId)
+	var authentications []model.Authentication
+
+	switch resourceType {
+	case "Source":
+		sourceDao := dao.GetSourceDao(tenantId)
+		applications, endpoints, rhcConnections, source, err := sourceDao.DeleteCascade(resourceId)
+		if err != nil {
+			return fmt.Errorf(`could not completely delete the source: %s`, err)
+		}
+
+		// Raise events for the deleted applications.
+		var appIds []int64
+		for _, app := range applications {
+			appIds = append(appIds, app.ID)
+			err := RaiseEvent("Application.Delete", &app, headers)
+			if err != nil {
+				logging.Log.Errorf(`Event "Application.Delete" could not be raised for application %v: %s`, app.ToEvent(), err)
+			}
+		}
+
+		// Raise events for the deleted endpoints.
+		var endpointIds []int64
+		for _, endpoint := range endpoints {
+			endpointIds = append(endpointIds, endpoint.ID)
+			err := RaiseEvent("Endpoint.Delete", &endpoint, headers)
+			if err != nil {
+				logging.Log.Errorf(`Event "Endpoint.Delete" could not be raised for endpoint %v: %s`, endpoint.ToEvent(), err)
+			}
+		}
+
+		// Raise events for the deleted connections.
+		for _, connection := range rhcConnections {
+			err := RaiseEvent("RhcConnection.Delete", &connection, headers)
+			if err != nil {
+				logging.Log.Errorf(`Event "RhcConnection.Delete" could not be raised for rhcConnection %v: %s`, connection.ToEvent(), err)
+			}
+		}
+
+		// Raise an event for the source itself.
+		err = RaiseEvent("Source.Delete", source, headers)
+		if err != nil {
+			logging.Log.Errorf(`Event "Source.Delete" could not be raised for source %v: %s`, source.ToEvent(), err)
+		}
+
+		// Fetch all the authentications from the resources.
+		resourceTypes := []struct {
+			ResourceType string
+			ResourceIds  []int64
+		}{
+			{ResourceType: "Application", ResourceIds: appIds},
+			{ResourceType: "Endpoint", ResourceIds: endpointIds},
+			{ResourceType: "Source", ResourceIds: []int64{resourceId}},
+		}
+
+		for _, res := range resourceTypes {
+			auths, err := authenticationsDao.ListIdsForResource(res.ResourceType, res.ResourceIds)
+			if err != nil {
+				logging.Log.Errorf(`[resource_type: "%s"][resource_ids: "%v"] Could not fetch authentications: %s`, res.ResourceType, res.ResourceIds, err)
+			}
+
+			authentications = append(authentications, auths...)
+		}
+
+	case "Application":
```

Comment:
We do need to delete the `ApplicationAuthentications` as well - which might need to be deleted first due to fkeys.

Reply:
I cascade deleted the applications too, to be able to notify about the deleted application authentications as well.